src/Slist/Maybe.hs

```diff
@@ -0,0 +1,130 @@
+{- |
+Copyright:  (c) 2021 Kowainik
+SPDX-License-Identifier: MPL-2.0
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+Stability:   Stable
+Portability: Portable
+
+Useful 'Maybe' combinators to work with the 'Maybe' data type and 'Slist'
+together.
+-}
+module Slist.Maybe
+    ( maybeToSlist
+    , slistToMaybe
+    , catMaybes
+    , mapMaybe
+    , slistWith
+    ) where
+
+import Slist.Type (Slist (..), cons, one)
+
+
+{- | Returns an empty list when given 'Nothing' or a singleton list when given
+'Just'.
+
+>>> maybeToSlist (Just 42)
+Slist {sList = [42], sSize = Size 1}
+>>> maybeToSlist Nothing
+Slist {sList = [], sSize = Size 0}
+-}
+maybeToSlist :: Maybe a -> Slist a
+maybeToSlist Nothing  = mempty
+maybeToSlist (Just x) = one x
+{-# INLINE maybeToSlist #-}
+
+{- | Returns 'Nothing' on an empty list or @'Just' a@ where @a@ is the first
+element of the slist.
+
+==== __Examples__
+
+Basic usage:
+
+>>> slistToMaybe mempty
+Nothing
+
+>>> slistToMaybe (one 42)
+Just 42
+
+>>> slistToMaybe (cons 1 $ cons 2 $ one 3)
+Just 1
+
+__Laws__ :
+
+@
+slistToMaybe . maybeToList = id
+@
+
+Reverse is right only on singleton/empty lists
+
+@
+maybeToList . slistToMaybe {empty, singleton slist} = id
+@
+-}
+slistToMaybe :: Slist a -> Maybe a
+slistToMaybe = foldr (const . Just) Nothing
+{-# INLINE slistToMaybe #-}
+-- We define listToMaybe using foldr so that it can fuse via the foldr/build
+-- rule. See #14387
+
+
+{- | Takes a slist of 'Maybe's and returns a slist of all the 'Just' values.
+
+>>> catMaybes (cons (Just 1) $ cons Nothing $ one $ Just 3)
+Slist {sList = [1,3], sSize = Size 2}
+-}
+catMaybes :: Slist (Maybe a) -> Slist a
+catMaybes = mapMaybe id
+{-# INLINE catMaybes #-}
+
+{- | The 'Maybe' version of 'map' which can throw out elements.
+
+If appliying the given function returns 'Nothing', no element is added on to the
+result list. If it is @'Just' b@, then @b@ is included in the result list.
+
+>>> maybeEven x = if even x then Just x else Nothing
+>>> s = cons 1 $ cons 2 $ one 3
+
+>>> mapMaybe maybeEven s
+Slist {sList = [2], sSize = Size 1}
+
+If we map the 'Just' constructor, the entire list should be returned:
+
+>>> mapMaybe Just s
+Slist {sList = [1,2,3], sSize = Size 3}
+-}
+mapMaybe :: forall b a . (a -> Maybe b) -> Slist a -> Slist b
+mapMaybe _ (Slist [] _) = mempty
+mapMaybe f (Slist (x:xs) n) = case f x of
+    Nothing -> rest
+    Just r  -> cons r rest
+  where
+    rest :: Slist b
+    rest = mapMaybe f (Slist xs (n - 1))
+{-# NOINLINE [1] mapMaybe #-}
+
+{- | Similar to 'mapMaybe' but works with the ordinary list as the input:
+
+>>> maybeEven x = if even x then Just x else Nothing
+
+>>> slistWith maybeEven [1,2,3]
+Slist {sList = [2], sSize = Size 1}
+-}
+slistWith :: forall b a . (a -> Maybe b) -> [a] -> Slist b
+slistWith _ [] = mempty
+slistWith f (x:xs) = case f x of
+    Nothing -> rest
+    Just r  -> cons r rest
+  where
+    rest :: Slist b
+    rest = slistWith f xs
```

Comment:
This implementation looks fine to me 👍🏻 
I'm thinking, if we can improve it. I see, that our `Size` field is lazy in the `Slist` type intentionally. And `cons` increments the size by one. So I'm afraid that we will have space leaks here.

I see that it's very difficult to solve this problem. If we returned only a list, we could just prepend the element, but we also need to track the size somehow.

So, ideally, I want to write a code like this:

```haskell
go !curSize list [] = Slist list curSize
go !curSize list (x:xs) = case f x of
    Nothing -> go curSize list xs
    Just r -> go (curSize + 1) (list ++ [x]) xs
```

But you see the obvious problem that appending to the end of the list is slow...

One option is to return a tuple and use `second (x:) <$> go curSize xs` to append lazily at the beginning of the list. This approach will allow us to track the `curSize` explicitly in `go` with explicit BangPattern.

The second option is to patch slightly our initial implementation, we can use DList (different list) instead of ordinary list. Because adding to the end of the different list is quite fast.

This problem actually always worried me. If you're working only with lists, it's quite straightforward to append to the beginning and write _streaming_ solutions. But once you need to return something else, you now need to do something weird... I'm also curious about performance and memory usage of various solutions. Would be a nice idea for a small blog post, comparing several solutions, providing some stats and explaining the results 🙂